opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java

```diff
@@ -706,8 +706,10 @@ public boolean runAutomation() throws SQLException {
         if (hasTrigger()) {
             LOG.debug("runAutomation: {} trigger statement is: {}", m_automation.getName(), m_trigger.getTriggerSQL());
         }
-            
-        LOG.debug("runAutomation: {} action statement is: {}", m_automation.getName(), m_action.getActionSQL());
+
+        if(hasAction()){
```

Comment:
This is a bit suspect, is it really legal for something called an `ActionProcessor` to have a `null` action? Under what circumstance is that allowed?

I realize it has a method to check if it has an action, but it looks like you are the first user of that and the fact it was present smells a little hacky to begin with.

Reply:
As far as I can find out, the reason ActionProcessor can have a null action is when the user input an "action name" that isn't in the action list, which could be caused by a typo by user.

Reply:
I think at the point, we might need to decide how to handle this senario. Do we just skip it or do we at least log this so customer know there is something wrong or do we really want to throw an exception?

Reply:
Ok so as far as I can tell, an `AutomationProcessor` delegates to an `ActionProcessor`. We added null checking to the `ActionProcessor` to guard against a null `Action` which probably doesn't make sense.

I'm tempted to say we should instead move the check up a level.

We could say that it is illegal for an `ActionProcessor` to have a `null` `Action`. Then we can guard against that via asserting an `IllegalArgumentException` in the `ActionProcessor` class.

Now in the constructor for `AutomationProcessor`, rather than set it's `ActionProcessor` to one with a null `Action`, we set the `ActionProcessor` instance to null itself.

Now you can add new guards in the processing logic to see if you have a non-null `ActionProcessor` or not and take action accordingly. Could also update the getter to return an `Optional` also (although the getter doesn't seem to be used anymore).

Does this make sense?

Reply:
Should be able to back out most of the changes made and also those in https://github.com/OpenNMS/opennms/pull/3285 if we follow through with this.

Reply:
Sure, just one second.

Reply:
Additionally, if we think it is an error case we could blow up the constructor of `AutomationProcessor` and log a warning rather than set its delegate to `null` and guard against it. I'm not familiar enough with the problem to know which is the right solution.

Reply:
I don't like null, and don't like assert. So I pick the second option and log the warning : -).

Reply:
What if we propagate the error up another level to the `AutomationProcessor` constructor. Can we just blow up there if it finds a null? Then we don't need any of this guard logic anymore.

Reply:
Either way would work. I think these are just different coding styles. the reason I don't like to put the null check in  AutomationProcessor constructor is that I think it doesn't hurt to have an AutomationProcessor with an "Action" that has an empty field "action". I just don't like to blow up the whole thing just because this null "action" doesn't make sense. Also throwing exceptions may impact performance, and this is automation keeps on running. Also, I want to log the warning to let the user know what's wrong and how they can fix it, put the log too deep in the code make the warning seems not too related to that place.   
Of course, there will also be a bright side to handling the error in AutomationProcessor constructor.  
Either way works for me. If we feel put the checking inside constructor more fits OpenNMS style, I have no problem doing that.

Reply:
By the way, I'm not sure what's our plan for Vacuumd, I guess we definitely will need to redesign it to make it user-friendly but not letting the user write SQL :P. Also configuring via XML is not desirable :D

Reply:
In this case I don't think it's just a style issue. I think the question is whether an `AutomationProcessor` is even valid in the case where we couldn't find an `Action` for it. I suspect the answer is now. In that case we want to fail as early as possible so we can accurately describe what went wrong and where.

Here is a patch for what I had in mind: https://github.com/OpenNMS/opennms/compare/mbrooks/null_action

Please take a look at my proposed change and let me know what you think.

This "fixes" the NPE by turning it into a proper error. I think the other attempts we were trying here to fix it were addressing the problem at too low a level and didn't really address the underlying problem.

With that out of the way, I think there is a separate issue here since an `Action`'s `Statement` can have empty content, does that make sense? Calling `getActionSQL()` can return an empty String which is a bit odd, it could probably be changed to an `Optional<String>` instead. Might not be worth addressing in this PR though as I think it is a separate issue.

Reply:
Please note I am making an assumption here about how an `AutomationProcessor` works and what its invariants are based on little understanding of the domain here so correct me if I'm wrong anyone else on this review.

Reply:
Happy Monday Matthew! The patch you share works for me too. Just one note, we probably don't need to check if the "actionname" is null, because the application will fail to start if we leave "actionname" blank.  I have another idea, how about we throw the exception  in where the null come from.  I'm attaching it here in the comment.
    public synchronized Action getAction(String actionName) {
        for (Action act : getActions()) {
            if (act.getName().equals(actionName)) {
                return act;
            }
        }
//        return null;
        throw new IllegalArgumentException("Could not find an action for automation action named '" + actionName + "'");
    }

If the requirement is user must have the "action" and it has to be valid, then this is the only code change we need to make.

Reply:
Hi Jane, good idea addressing the issue there, however I think we could improve it even a bit further. Rather than throw an exception in `getAction` we can simply return a `Optional<Action>` and leave the exception throwing up to the caller still. This allows the API of `getAction` to remain flexible.

So our `AutomationProcessor` constructor would look like this:
```
        String actionName = automation.getActionName();
        Action actionForAutomation = VacuumdConfigFactory.getInstance()
                .getAction(actionName)
                .orElseThrow(() -> new IllegalArgumentException("Could not find an action for automation action named '" + actionName + "'"));
```

I think with this solution we will be good to go.

Reply:
Hi Matthew. I like this idea! I've updated the code and tested on my machine.

Reply:
Thanks Jane, looks good.